### DISCUSSION TITLE
crypto.randomUUID is not a function

### DISCUSSION THREAD

**Discussion by jladbury** [rejetto/hfs#124] [Q&A] [answered]
At first I was going to raise this as a bug, as it does not occur in HFS 2.3, nor in the WebStorm server. However, a bit of investigation shows that somehow my application is involved.

The application makes a call to `crypto.randomUUID`. As I say, this works fine in HFS 2.3 and in the WebStorm server, but fails in HFS 3.

I searched the repository for 'crypto' and found a recent change, and upgraded to the latest version (0.37.0). The problem remains.

The strange thing is that if I use devtools to enter `crypto.randomUUID` in the console BEFORE loading my app, it works fine. However, after the app is loaded I get the error despite not making any (intentional) change to crypto-related stuff.

I would be grateful for any suggestions, as I want to move permanently to the new HFS; but for the moment this will stop me.

**Comment by rejetto** [comment]
hi Julian, what do you mean by "my application" ?
did you make a web app, and are using hfs just to serve its static files?

**Reply by jladbury** [reply]
Yes indeed!

**Comment by rejetto** [comment, marked as answer]
if I had to guess, you are missing crypto because you are neither on localhost or in https.
If this is it, then it has nothing to do with hfs 3.

**Reply by jladbury** [reply]
That's definitely something to do with it. I will document further and request your comments.
By the way, many thanks for such a useful tool.

**Reply by jladbury** [reply]
Yes, that's nailed it!
In HFS 3:

- Access from `http://192.168.1.180/c9/c9define/compiled/index.htm` gives the problem
- Access from `http://localhost/c9/c9define/compiled/index.htm` works fine

As to HFS 2, it defaults to `localhost` if I press 'Open in Browser'. However, if I copy and paste the URL it displays, I get the problem. So, as you suggest, this has nothing to do with HFS version.

Might I suggest that the instructions at the top of the Shared Files page in the Admin panel (`Your shared files can be browsed from http://192.168.1.180/ or http://192.168.56.1/ or http://[fe80::dd32:bd01:25cb:f1b8]/` be updated to include `localhost`?

Also, if you have time to explain briefly why the `localhost` differs from the actual IP address I would be grateful!

**Reply by rejetto** [reply]
thanks for your suggestion, but I'm not very sure including localhost in the list would cause more good than evil.
localhost doesn't work over the network, and that's how i expect people to use it. It says "shared files" for this reason.
While the old "open in browser" button wasn't meant to share, but just to open your browser, so localhost was fine.

your problem with crypto was caused by a decision made by web-browsers in recent years. I don't know their reasons, but you may google for it. 
I had to fight with it as well 🫠

**Reply by jladbury** [reply]
Thanks for taking the time to comment. Just FYI, my use case for HFS is local testing rather than file sharing.

**Reply by rejetto** [reply]
yes, I see that, and I'm glad to see people find it useful for other purposes as well 😊
